### PR TITLE
Add shorthands to get features by fid, fids, rectangle and expression

### PR DESCRIPTION
**Pull request by m-kuhn**
Adds new methods overloads to QgsVectorLayer

```
  getFeatures( expression )
  getFeatures( ids )
  getFeature( id )
```

These three methods to query features are by far the most used ones
and with this patch it is much easier to write and read code.

### CODE REVIEW

**Comment by mbernasocchi** [comment]
+100
Thanks a lot

**Comment by wonder-sk** [comment]
If we add these variants for convenience, how about:
- they are added also in QgsVectorDataProvider and QgsVectorLayerCache for consistency
- the variant for QgsRectangle is also added
- they are inline functions

**Comment by m-kuhn** [comment]
+1 for QgsVectorLayerCache

-0 for QgsVectorDataProvider. Reasoning: I think a lot of people are not aware of the implications of directly using the dataprovider methods and just do it because they have seen or copied it from somewhere. So making it a bit harder to use it is an easy way to discourage usage without prohibiting it.
+1 for rectangle
+1 for inline

**Comment by nyalldawson** [comment]
Also, some basic unit tests would be good...

> they are inline functions

Is there any point to marking stuff inline anymore?    I was of the opinion the modern compilers just ignore it and make their own choices re inclining anyway...

File: src/core/qgsvectorlayer.h [hunk 1]
```diff
@@ -985,6 +985,22 @@ class CORE_EXPORT QgsVectorLayer : public QgsMapLayer
...
+    /**
+     * Query the provider for the feature with the given id.
+     * If there is no such feature, the returned feature will be invalid.
+     */
+    QgsFeature getFeature( QgsFeatureId fid );
+
+    /**
+     * Query the provider for the features with the given ids.
+     */
+    QgsFeatureIterator getFeatures( QgsFeatureIds fids );
```
Comment (nyalldawson) [inline src/core/qgsvectorlayer.h]: Should fids be a const ref here? I realise it's shared, but I kind of like making parameters const refs as its a bit more explicit with no risk, plus it follows the qt style

Comment (m-kuhn) [inline src/core/qgsvectorlayer.h]: Right

**Comment by m-kuhn** [comment]
Unit tests are on my list, just wanted to have the general ok first :)

**Comment by nyalldawson** [comment]
Ok, +1 from me too then!

**Comment by wonder-sk** [comment]
> -0 for QgsVectorDataProvider. Reasoning: I think a lot of people are not aware of the implications of directly using the dataprovider methods and just do it because they have seen or copied it from somewhere. So making it a bit harder to use it is an easy way to discourage usage without prohibiting it.

I would say it should be the same everywhere - otherwise it will break user's expectations. If there will be more classes with getFeatures() later, how do we tell which ones get the convenience methods and which ones do not? I think the issue with people using data providers directly is better addressed in documentation and code samples...

> Is there any point to marking stuff inline anymore? I was of the opinion the modern compilers just ignore it and make their own choices re inclining anyway...

I was mainly referring to the fact that the implementation could go to the header file (inside or outside the class), just like Qt does with various convenience methods. The "inline" keyword is indeed mostly ignored as far as I know.

**Comment by m-kuhn** [comment]
> I would say it should be the same everywhere - otherwise it will break user's expectations.

dataprovider methods break the users expectations more often than they are useful, I am sure.

> If there will be more classes with getFeatures() later, how do we tell which ones get the convenience methods and which ones do not?

Predicting the future is something we should better leave to fortune-tellers

> I think the issue with people using data providers directly is better addressed in documentation and code samples...

Yes we should. But who reads that anyway. And how are we going to fix the tons of samples on gis.se?

**Comment by wonder-sk** [comment]
> dataprovider methods break the users expectations more often than they are useful, I am sure.

Shall we remove those dataprovider methods then? ;-) But seriously, not having other getFeatures() overrides is not going to increase "protection" of users from doing anything bad in data providers...

> Predicting the future is something we should better leave to fortune-tellers

I have real use cases in mind :-)

> Yes we should. But who reads that anyway. And how are we going to fix the tons of samples on gis.se?

gis.se samples can be of course fixed - by upvoting good ones and downvoting bad ones...

**Comment by m-kuhn** [comment]
> Shall we remove those dataprovider methods then? 

I don't know... Do you have real use cases for them in mind?

Over the years I've been fixing many bad usages in core C++ code and in python but I haven't found (m)any examples where it's required or even handy.

> But seriously, not having other getFeatures() overrides is not going to increase "protection" of users from doing anything bad in data providers...

I disagree, it does increase protection because people start using the methods which are only available on QgsVectorLayer. At the same time it's not going to hurt anyone.
The 2 legitimate users of the dataprovider API can just continue to use it with non-shorthands and if 2 of the 1000 coders that used it accidentally switched over we have already done something good :)

**Comment by wonder-sk** [comment]
Often it is simply more efficient to use directly data provider methods if you don't need all the extra stuff from vector layer. But I don't want to hijack this PR to discuss the topic of data provider VS vector layer...

One thing I thought about was to introduce feature source interface in QGIS API, so any class that implements its getFeatures() virtual method could be used in various cases where iterating over features is used, mainly for processing algorithms (no need to deal with full vector layers all the time). In such interface it would make sense to directly have those convenience functions to avoid duplicate code (and so data providers would get those too).

**Comment by m-kuhn** [comment]
> Often it is simply more efficient to use directly data provider methods if you don't need all the extra stuff from vector layer.

We should rather fix that ;) Seriously if the subsetOfAttributes doesn't reference virtual layers or joined fields and the layer is not in edit mode, there is really no reason for this method to be slower.

> One thing I thought about was to introduce feature source interface in QGIS API

Yes, I have thought about that as well. This would avoid things like signature mismatch (e.g. `pendingFields()` vs. `fields()` ) and allow a lot more generic programming. And in this case I wouldn't mind about the data providers also receiving these functions ;)

**Comment by m-kuhn** [comment]
Ok if I merge it without dataprovider methods for now? ;)

**Pull request closed without merging**